cosmos/job/drm/drm_ge.py

```diff
@@ -301,17 +301,88 @@ def qacct(job_id, num_retries=10, quantum=30, logger=None, log_prefix=""):
     return good_qacct_dict if good_qacct_dict else curr_qacct_dict
 
 
-def qstat():
+def qdel(job_ids, logger):
+    """
+    Call qdel on all the supplied job_ids: if that fails, qdel each job_id individually.
+
+    Unlike other SGE cli commands, each qdel call is attempted only once, with a
+    fairly harsh one-second timeout, because function method is often called in
+    an exit handler that does not have arbitrary amounts of time in which to run.
+    """
+    stdout, stderr, returncode = run_cli_cmd(
+        ["qdel", ",".join(job_ids)],
+        logger=logger,
+        attempts=1,
+        timeout=1,
+    )
+    if returncode == 0:
+        logger.info("qdel reported successful signalling of %d job_ids", len(job_ids))
+        return len(job_ids)
+
+    successful_qdels = 0
+    undead_job_ids = []
+
+    for job_id in job_ids:
+        if "has deleted job %s" % job_id in stdout:
+            successful_qdels += 1
+        elif "has registered the job %s for deletion" % job_id in stdout:
+            successful_qdels += 1
+        else:
+            undead_job_ids.append(job_id)
+
+    if undead_job_ids:
+        logger.warning(
+            "qdel returned exit code %s, calling on one job_id at a time", returncode
+        )
+
+        for job_id in undead_job_ids:
+            _, _, returncode = run_cli_cmd(
+                ["qdel", job_id], logger=logger, attempts=1, timeout=1,
+            )
+            if returncode != 0:
+                logger.warning(
+                    "qdel returned %s attempting to kill %s: it may still be running",
+                    returncode,
+                    job_id,
+                )
+            else:
+                successful_qdels += 1
+
+    logger.info(
+        "qdel reported successful signalling of %d of %d job_ids",
+        successful_qdels,
+        len(job_ids),
+    )
+    return successful_qdels
+
+
+def qstat(logger=None):
     """
     Return a mapping of job ids to a dict of GE information about each job.
 
+    If qstat hangs or returns an error, wait 30 sec and call it again. Do this
```

Comment:
this comment is false -- if hangs for more than 30 sec, wait 15 sec and retry. update run_cli_cmd invocation with all params to make this self-documenting & fix the comment